### PR TITLE
Conditional logging

### PR DESCRIPTION
**Pull request by MorrisJobke**
Log condition for log level increase based on conditions. Once one of these conditions is met, the required log level is set to debug. This allows to debug specific requests, users or apps

Supported conditions (setting `log_condition` in `config.php`):
    - `secret`: if a request parameter with the name `log_secret` is set to this value the condition is met
    - `users`:  if the current request is done by one of the specified users, this condition is met
    - `apps`:   if the log message is invoked by one of the specified apps, this condition is met
- [x] requires #15958 (which is included here to easiely test this)

This would be really helpful to debug bigger installations where enabling a higher log level isn't a option because of the huge amount of log messages. cc @felixboehm @butonic 

cc @nickvergessen @PVince81 @rullzer @Xenopathic @Raydiation @DeepDiver1975

### CODE REVIEW

**Comment by MorrisJobke** [comment]
* core commit here is 753c2db
* cleanup and code deduplication is in here: https://github.com/owncloud/core/commit/2dc650c92db76e44a20050441f8ad4b16d7cf6d3?w=1

**Comment by MorrisJobke** [comment]
I used the internal `\OC\SystemConfig` here to not run into cyclic conditions here. As this is the logger I want to reduce the needed bootstrapped system to a minimum. Otherwise it could be possible to no able to log, because the injected object couldn't be build. The system config was also required previously. And as this is an internal class I think this is not a problem. Opinions?

**Comment by felixboehm** [comment]
Awesome stuff, logfiles often are so big and this really helps

**Comment by karlitschek** [comment]
nice idea :+1:

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by DeepDiver1975** [comment]
rebased after merge of #15958

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by DeepDiver1975** [comment]
@MorrisJobke unit tests are broken


````
Test\Logger::testInterpolation
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => '1 {Message {nothing} Bob Bar a}'
 )

/var/jenkins/workspace/pull-request-analyser-ng-simple@3/label/SLAVE/tests/lib/logger.php:32

**Comment by MorrisJobke** [comment]
I added a unit test for the app condition.

**Comment by MorrisJobke** [comment]
@LukasReschke @nickvergessen @rullzer @Xenopathic Do you want to review?

**Comment by LukasReschke** [comment]
"want"

File: config/config.sample.php [hunk 1]
```diff
@@ -487,6 +487,27 @@
 'loglevel' => 2,
 
 /**
+ * Log condition for log level increase based on conditions. Once one of these
+ * conditions is met, the required log level is set to debug. This allows to
+ * debug specific requests, users or apps
+ *
+ * Supported conditions:
+ *  - ``secret``: if a request parameter with the name `log_secret` is set to
```
Comment (LukasReschke) [inline config/config.sample.php]: `shared_secret`?

File: config/config.sample.php [hunk 2]
```diff
@@ -487,6 +487,27 @@
...
+ *                this value the condition is met
+ *  - ``users``:  if the current request is done by one of the specified users,
+ *                this condition is met
+ *  - ``apps``:   if the log message is invoked by one of the specified apps,
+ *                this condition is met
+ *
+ * Defaults to an empty array.
+ */
+'log_condition' => [
```
Comment (LukasReschke) [inline config/config.sample.php]: I prefer something like `log.condition` :see_no_evil:

File: lib/private/log.php [hunk 3]
```diff
@@ -158,8 +172,22 @@ public function debug($message, array $context = array()) {
...
 		if (isset($context['app'])) {
 			$app = $context['app'];
+
+			/**
+			 * check log condition based on the context of each log message
+			 * once this is met -> change the required log level to debug
+			 */
+			if(!empty($logCondition)
+				&& isset($logCondition['apps'])
+				&& in_array($app, $logCondition['apps'])) {
```
Comment (LukasReschke) [inline lib/private/log.php]: Always use the third strict parameter for `in_array`. Thanks.

File: lib/private/log.php [hunk 4]
```diff
@@ -172,7 +200,45 @@ public function log($level, $message, array $context = array()) {
...
+			// default to false to just process this once per request
+			$this->logConditionSatisfied = false;
+			if(!empty($logCondition)) {
+
+				// check for secret token in the request
+				if(isset($logCondition['secret'])) {
+					$request = \OC::$server->getRequest();
+
+					// if token is found in the request change set the log condition to satisfied
+					if($request && $request->getParam('log_secret') === $logCondition['secret']) {
```
Comment (LukasReschke) [inline lib/private/log.php]: I'd prefer if we would use a header for something like this. Params seem rather error prone to me. `\OCP\IRequest::getHeader`

Comment (LukasReschke) [inline lib/private/log.php]: Also for security purposes to ensure a constant time comparison use \OC\Security\StringUtils::equals

Comment (butonic) [inline lib/private/log.php]: I would prefer if it was possible to combine app and user filers: only log file cache operations for user a.

Am 8. Mai 2015 13:58:49 MESZ, schrieb Lukas Reschke <notifications@github.com>:
>> -		call_user_func(array($logger, 'write'), $app, $message, $level);
>> +		/**
>> +		 * check for a special log condition - this enables an increased
>log on
>> +		 * a per request/user base
>> +		 */
>> +		if($this->logConditionSatisfied === null) {
>> +			// default to false to just process this once per request
>> +			$this->logConditionSatisfied = false;
>> +			if(!empty($logCondition)) {
>> +
>> +				// check for secret token in the request
>> +				if(isset($logCondition['secret'])) {
>> +					$request = \OC::$server->getRequest();
>> +
>> +					// if token is found in the request change set the log
>condition to satisfied
>> +					if($request && $request->getParam('log_secret') ===
>$logCondition['secret']) {
>
>Also for security purposes to ensure a constant time comparison use
>\OC\Security\StringUtils::equals
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/owncloud/core/pull/15965/files#r29933978

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

Comment (DeepDiver1975) [inline lib/private/log.php]: > I'd prefer if we would use a header for something like this. Params seem rather error prone to me. \OCP\IRequest::getHeader

I guess the idea is to append this parameter to any page load in the browser - not that easy if we use headers ....

Comment (MorrisJobke) [inline lib/private/log.php]: > I would prefer if it was possible to combine app and user filers: only log file cache operations for user a.

We either decide on "AND" or "OR" ... otherwise it will get very complex.

Comment (LukasReschke) [inline lib/private/log.php]: > I guess the idea is to append this parameter to any page load in the browser - not that easy if we use headers ....

One shall use an interception proxy for this :see_no_evil: 


/me hides again ;-)

File: lib/private/log.php [hunk 5]
```diff
@@ -172,7 +200,45 @@ public function log($level, $message, array $context = array()) {
...
+						$this->logConditionSatisfied = true;
+					}
+				}
+
+				// check for user
+				if(isset($logCondition['users'])) {
+					$user = \OC::$server->getUserSession()->getUser();
+
+					// if the user matches set the log condition to satisfied
+					if($user !== null && in_array($user->getUID(), $logCondition['users'])) {
```
Comment (LukasReschke) [inline lib/private/log.php]: `in_array` always with strict parameter. Thanks :)

File: lib/private/log/owncloud.php [hunk 6]
```diff
@@ -69,40 +69,40 @@ public static function init() {
...
+		$time = new DateTime(null, $timezone);
+		$request = \OC::$server->getRequest();
+		$reqId = $request->getId();
+		$remoteAddr = $request->getRemoteAddress();
+		// remove username/passwords from URLs before writing the to the log file
+		$time = $time->format($format);
+		$minLevel=min($config->getValue( "loglevel", OC_Log::WARN ), OC_Log::ERROR);
+		if($minLevel == OC_Log::DEBUG) {
+			$url = isset($_SERVER['REQUEST_URI']) ? $_SERVER['REQUEST_URI'] : '--';
+			$method = isset($_SERVER['REQUEST_METHOD']) ? $_SERVER['REQUEST_METHOD'] : '--';
```
Comment (LukasReschke) [inline lib/private/log/owncloud.php]: While already at it can we adjust this to the IRequest methods as well?

Comment (MorrisJobke) [inline lib/private/log/owncloud.php]: I just touched the intendation here. I would like to not touch the logic in here even more.

Comment (LukasReschke) [inline lib/private/log/owncloud.php]: Lazy slacker! ;-)

**Comment by MorrisJobke** [comment]
Do we want to use a header or not?

**Comment by MorrisJobke** [comment]
BTW: all the other comments are addressed.

File: config/config.sample.php [hunk 7]
```diff
@@ -487,6 +487,27 @@
...
+ *                this value the condition is met
+ *  - ``users``:  if the current request is done by one of the specified users,
+ *                this condition is met
+ *  - ``apps``:   if the log message is invoked by one of the specified apps,
+ *                this condition is met
+ *
+ * Defaults to an empty array.
+ */
+'log.condition' => [
+	'secret' => '57b58edb6637fe3059b3595cf9c41b9',
```
Comment (LukasReschke) [inline config/config.sample.php]: Erm? :see_no_evil:

Comment (LukasReschke) [inline config/config.sample.php]: Now doc and content is not the same anymore :see_no_evil:

Comment (MorrisJobke) [inline config/config.sample.php]: Oh ... forgot that one

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by MorrisJobke** [comment]
@LukasReschke @nickvergessen @rullzer @Xenopathic Ready for review :)

**Comment by rullzer** [comment]
Working very nice!
I would like to see a header as well (maybe different PR, if it is to much work since this should get in). So debugging the OCS API etc becomes a bit easier.

As mentioned on IRC the call_user_func stuff makes me cry. But also that should be addressed in another PR.

:+1: !

File: lib/private/log/errorlog.php [hunk 8]
```diff
@@ -39,10 +39,7 @@ public static function init() {
 	 * @param int $level
 	 */
 	public static function write($app, $message, $level) {
-		$minLevel = min(OC_Config::getValue("loglevel", OC_Log::WARN), OC_Log::ERROR);
-		if ($level >= $minLevel) {
```
Comment (DeepDiver1975) [inline lib/private/log/errorlog.php]: why was this removed? @MorrisJobke THX

**Comment by DeepDiver1975** [comment]
:+1:
